fix: native messages layer (part 1)

BrianAmadori opened the pull request:
### Introduction

This is the first part in a series of performance improvements aimed at optimizing the messaging between kernel and unity. The first objective is to remove the hiccups the following overheads, measured from profiling key scenes:

* Protobuf's `ParseFrom`
* `string.Split` from `sendBatch` methods
* base64 encoding/decoding

These overheads are going to be removed by _stripping protobuf entirely_ from the wasm messaging pipeline, and introducing a new direct call pipeline, that will copy values directly through wasm's heap and direct calls to emscripten generated methods.

Protobuf isn't going to be removed because is still useful for the websocket-powered debugging. Also, it will still prevail when we go for the hybrid native client.

### Proofs (with .data files included to check by yourself in unity profiler)
With street cred to @pravusjif and @sandrade-dcl 

https://docs.google.com/document/d/1v7OTZywBfbssqBusboidlE29jRYM0xlq0S-SQt6jyZY/edit
https://docs.google.com/document/d/1O_RSYMk3a4qDE-481dL2RrBrCI6CYjjx4eS7X_DQKMQ/edit
https://docs.google.com/document/d/1Fkw73nY0t4y1u2KnjWOONF_sT4D_cifToGzXpurJndU/edit
https://docs.google.com/document/d/1s6to9chPTqxAUTj0ZxgjbHiP5OV-qlXN3jdnHUTo_vE/edit

### Direct calls vs protobuf measurements
https://docs.google.com/document/d/1-0ja-_F-sjdUhldSfIDHWcGjtAMDC83M6_NtO5QPL_c/edit
https://github.com/BrianAmadori/UnityJavascriptInteraction/pull/2


----
### Test links

Vanilla:
https://play.decentraland.zone/branch/fix/native-messages-layer-part-1/?ENV=org

### Benchmark links
To benchmark properly, you must move around until the messages count gets up to 10000. Benchmark info will output on devtools console.

Direct calls:
https://play.decentraland.zone/branch/fix/native-messages-layer-part-1/?ENV=org&DEBUG_MESSAGES_QUEUE_PERF

SendMessage:
https://play.decentraland.zone/branch/fix/native-messages-layer-part-1/?ENV=org&DEBUG_MESSAGES_QUEUE_PERF&FORCE_SEND_MESSAGE

----
### Benchmark results:

All tests made with chrome v84 (64 bits), walking around genesis plaza:

In i7-6700HQ CPU @2.60Ghz + GTX 960m
direct calls: 10000 messages = ~300ms
SendMessage: 10000 messages = ~1100ms

In i7-9750H CPU @2.60Ghz + GTX 1650 
direct calls: 10000 messages = ~220ms
SendMessage: 10000 messages = ~550ms

### Optimization roadmap

- [X] Part 1: Implement direct calls for all scene messages (ECS and raycast queries) <- you are here
- [ ] Part 2: Implement direct calls for LoadParcelScenes/UpdateParcelScenes/UnloadParcelScenes messages. Optimize ParcelScenes messages payload.
- [ ] Part 3: Drop json for component payloads, copy the payloads directly throught the wasm heap (and only update the fields that changed!)
- [ ] Part 4: pass entityIds/componentIds as integers instead base36 strings (as they are marshalled way faster into wasm's heap)

Commit:
native messages layer first draft

Commit:
fix: wss connection related errors

Commit:
first working version

Commit:
add new messages support draft

Commit:
all messages working except Query

Commit:
query integration first draft

Commit:
query fixes

- kernel fix: query ids are generated with base 10 ints, so they are compatible with int-casting and can be passed easier through the heap block

- unity fix: moved protobuf query payload decoding away from ProcessMessages, so we keep the compatibility with direct calls

Commit:
marge with master

Commit:
refactor: solved cyclic references by splitting dcl.ts in several files

- move unityInterface to its own file and is now a typed class
- move browserInterface to its own file and is now a typed class
- a bunch of linter fixes

Commit:
refactor: solved cyclic references by splitting dcl.ts in several files

- move unityInterface to its own file and is now a typed class
- move browserInterface to its own file and is now a typed class
- a bunch of linter fixes

Commit:
fix: undefined property issue on kernel init

Commit:
fix: raycastType not being passed properly

- refactor: move all raycast related stuff to the same folder, also moved Protocol.cs to a new assembly

Commit:
fix: many message enqueue micro-optimizations

- improved performance on the following: string.op_Equality, singleton getters, dictionary get,

Commit:
fix: query callbacks not working

Commit:
fix: add kernel messages performance flag

Commit:
fix lint issues

Commit:
Merge remote-tracking branch 'origin/master' into fix/native-messages-layer-part-1

Commit:
merge with master

 - solved many conflicts in dcl.ts, moved hotScenes stuff to unityInterface and browserInterface

Commit:
linter fixes

sandrade-dcl reviewed (approved):
All code reviewed and several tests done. I can feel the improvement during the scenes loading process (and the metrics in the console log are evident). Anyway I think we should do a more detailed profiling comparing between prod and this PR in order to know how it has improved. Good job!!

Commit:
revert helpers.ts changes

Commit:
add native calls integration guide

Commit:
typo

Commit:
Merge branch 'master' into fix/native-messages-layer-part-1

Commit:
some minor method renames

Commit:
fix: explorer HUDs showing in preview and builder when it shouldn't

Commit:
lint fixes

Commit:
fix: base10 to base36 queryId conversion was not working.

Changed the approach to map string-based ids to unique ints entirely from scene worker side.

Commit:
Merge branch 'master' into fix/native-messages-layer-part-1
